src/matUtils/uncertainty.cpp

```diff
@@ -0,0 +1,357 @@
+    return distvs;
+}
+
+size_t get_neighborhood_size(std::vector<MAT::Node*> nodes, MAT::Tree* T) {
+    //first step for this is collecting the full paths back to the root for all nodes
+    assert (nodes.size() > 1); //doesn't make sense if there's only one best placement.
+    std::vector<std::vector<MAT::Node*>> parentvecs;
+    for (size_t s=0; s<nodes.size(); s++) {
+        if (!nodes[s]->is_root()){ //if one of the epps sites is directly off the root, then this doesn't make much sense
+            std::vector<MAT::Node*> npath;
+            npath.emplace_back(nodes[s]); //include the node itself on the path so branch length is correctly accessed
+            for (auto& it: T->rsearch(nodes[s]->identifier)) {
+                npath.emplace_back(it);
+            }
+            parentvecs.emplace_back(npath);
+        } else { //instead, just construct a path of length 1 that contains the root node only
+            std::vector<MAT::Node*> npath;
+            npath.emplace_back(nodes[s]);
+            parentvecs.emplace_back(npath);
+        }
+    }
+    //then we need to identify all common elements to all node vectors
+    std::vector<MAT::Node*> common_nodes = get_common_nodes(parentvecs);
+    assert (common_nodes.size() > 0); //bare minimum this will always include the root. therefore it is always > 0
+    //then for all common nodes, we need to calculate the largest sum of paired distances for all samples to that specific common ancestor
+    //the smallest of these largest sums is the best neighborhood size value
+    size_t best_size = T->get_parsimony_score(); //bigger than the biggest maximum limit on neighborhood size. basically a big number
+    for (size_t s=0; s<common_nodes.size(); s++) {
+        //get the set of distances between each placement to this common ancestor with the path vectors
+        std::vector<float> distances = get_all_distances(common_nodes[s], parentvecs);
+        // for (auto i = distances.begin(); i != distances.end(); ++i) {
+        //     std::cerr << *i << ',';
+        // }
+        // //now find the biggest sum of shared values for this specific common node
+        float widest = 0.0;
+        for (size_t i=0; i<distances.size(); i++) {
+            for (size_t j=0; j<distances.size(); j++) {
+                if (i!=j){
+                    float spairdist = distances[i] + distances[j];
+                    if (spairdist > widest){
+                        widest = spairdist;
+                    }
+                }
+            }
+        }
+        //after that oddness, I now have a value which is the longest path going between any two nodes in the placement set
+        //which goes through this specific common ancestor
+        //admittedly this is probably not the fastest way to do this, I should be able to eliminate common ancestors which are directly ancestral to common ancestors between the same complete set without counting them up
+        //but I'm focusing on results first here

```

Comment:
@jmcbroome you should probably use an editor which adds a newline automatically when the comment gets really long.